### PR TITLE
Make the JavaParserTypeSolver store the path in the compilation units…

### LINKED ISSUE #1552: Can't resolve path of method declaration
I'm using JavaParser and JavaSymbolSolver to extract the method calls from a Java class file.
I'm interested in extracting, for each of the method calls within the class:

1. the line range of the method call
2. the qualified signature of the called method
3. the line range of the called method's declaration
4. the path of the java class containing the called method declaration

So far I had success with 1-3 but not so much with 4.
Here is the code that I'm using:

```java
package nl.tud.mcextractor;

import com.github.javaparser.JavaParser;
import com.github.javaparser.ast.CompilationUnit;
import com.github.javaparser.ast.Node;
import com.github.javaparser.ast.body.MethodDeclaration;
import com.github.javaparser.ast.expr.MethodCallExpr;
import com.github.javaparser.resolution.declarations.ResolvedMethodDeclaration;
import com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade;
import com.github.javaparser.symbolsolver.javaparsermodel.declarations.JavaParserMethodDeclaration;
import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
import com.github.javaparser.symbolsolver.resolution.typesolvers.CombinedTypeSolver;
import com.github.javaparser.symbolsolver.resolution.typesolvers.JavaParserTypeSolver;
import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeSolver;

import java.io.File;
import java.util.List;

public class Runner {
    private static final String FILE_PATH = "jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java";

    public static void main(String[] args) throws Exception {
        TypeSolver typeSolver = new CombinedTypeSolver(
                new ReflectionTypeSolver(),
                new JavaParserTypeSolver(new File("jNumerus/src/main/java")));

        CompilationUnit cu = JavaParser.parse(new File(FILE_PATH));

        System.out.format("Analyzing method calls of file %s\n", cu.getStorage().get().getPath());

        List<MethodCallExpr> methodCalls = cu.findAll(MethodCallExpr.class);
        for (MethodCallExpr mc : methodCalls) {
            ResolvedMethodDeclaration correspondingDeclaration = JavaParserFacade.get(typeSolver).solve(mc).getCorrespondingDeclaration();
            if (correspondingDeclaration instanceof JavaParserMethodDeclaration) {
                JavaParserMethodDeclaration declaration = (JavaParserMethodDeclaration) correspondingDeclaration;
                Node wrappedNode = declaration.getWrappedNode();
                MethodDeclaration wrappedDeclaration = (MethodDeclaration) wrappedNode;
                System.out.format("L%dC%d,L%dC%d -> %s, (L%dC%d,L%dC%d)\n",
                        mc.getBegin().get().line,
                        mc.getBegin().get().column,
                        mc.getEnd().get().line,
                        mc.getEnd().get().column,
                        correspondingDeclaration.getQualifiedSignature(),
                        wrappedDeclaration.getBegin().get().line,
                        wrappedDeclaration.getBegin().get().column,
                        wrappedDeclaration.getEnd().get().line,
                        wrappedDeclaration.getEnd().get().column);
                Node temp = wrappedNode.findRootNode();
                if (temp instanceof CompilationUnit) {
                    CompilationUnit compilationUnit = (CompilationUnit) temp;
                    if (compilationUnit.getStorage().isPresent()) {
                        System.out.format("Path: %s\n", compilationUnit.getStorage().get().getPath());
                    } else {
                        System.out.println("Couldn't resolve path");
                    }
                }
            }
        }
    }
}
```
the sample project that I'm trying to analyze is this one https://github.com/TheMatjaz/jNumerus.
This is the output:
```
Analyzing method calls of file mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java

L98C9,L98C33 -> it.matjaz.jnumerus.RomanInteger.setValueAndNumeral(int), (L122C5,L125C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L110C9,L110C35 -> it.matjaz.jnumerus.RomanInteger.setNumeralAndValue(it.matjaz.jnumerus.RomanNumeral), (L136C5,L139C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L123C24,L123C61 -> it.matjaz.jnumerus.RomanConverter.integerToRomanNumeral(int), (L186C5,L194C5)
Couldn't resolve path
L137C22,L137C61 -> it.matjaz.jnumerus.RomanConverter.romanNumeralToInteger(it.matjaz.jnumerus.RomanNumeral), (L122C5,L124C5)
Couldn't resolve path
L170C9,L170C33 -> it.matjaz.jnumerus.RomanInteger.setValueAndNumeral(int), (L122C5,L125C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L182C9,L182C35 -> it.matjaz.jnumerus.RomanInteger.setNumeralAndValue(it.matjaz.jnumerus.RomanNumeral), (L136C5,L139C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L223C27,L223C42 -> it.matjaz.jnumerus.RomanInteger.getValue(), (L146C5,L148C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L226C16,L226C54 -> it.matjaz.jnumerus.RomanNumeral.equals(java.lang.Object), (L328C5,L338C5)
Couldn't resolve path
L226C36,L226C53 -> it.matjaz.jnumerus.RomanInteger.getNumeral(), (L155C5,L157C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L239C30,L239C47 -> it.matjaz.jnumerus.RomanNumeral.toString(), (L352C5,L355C5)
Couldn't resolve path
L267C16,L267C25 -> it.matjaz.jnumerus.RomanInteger.getValue(), (L146C5,L148C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L280C16,L280C25 -> it.matjaz.jnumerus.RomanInteger.getValue(), (L146C5,L148C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L293C16,L293C25 -> it.matjaz.jnumerus.RomanInteger.getValue(), (L146C5,L148C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L306C16,L306C25 -> it.matjaz.jnumerus.RomanInteger.getValue(), (L146C5,L148C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L326C13,L326C27 -> it.matjaz.jnumerus.RomanInteger.intValue(), (L265C5,L268C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L328C20,L328C34 -> it.matjaz.jnumerus.RomanInteger.intValue(), (L265C5,L268C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
```
Basically I can extract the path only when the called method is declared in the class that I'm analyzing itself. Any idea on how to extract it also for methods declared in other parts of the codebase?

### LINKED ISSUE #1576: JavaParserTypeSolver should be configurable with ParserConfiguration
By modifying the configuration you could tell the parser to skip features you don't need, which will speed up the solving process.

### PR DESCRIPTION
**Pull request by matozoid**
Fixes #1552 
Fixes #1576

### CODE REVIEW

**Comment by matozoid** [comment]
I really shouldn't open PR's without building locally...

File: javaparser-core/src/main/java/com/github/javaparser/JavaParser.java [hunk 1]
```diff
@@ -330,7 +330,7 @@ public static Statement parseStatement(final String statement) {
         if (result.isSuccessful()) {
             return result.getResult().get();
         }
-        throw new ParseProblemException(result.getProblems());
+        throw new ParseProblemException(result);
```
Comment (ftomassetti) [inline javaparser-core/src/main/java/com/github/javaparser/JavaParser.java]: Why are the problems not enough?

File: javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/declarations/JavaParserInterfaceDeclaration.java [hunk 2]
```diff
@@ -310,11 +310,12 @@ public AccessSpecifier accessSpecifier() {
 
     private ResolvedReferenceType toReferenceType(ClassOrInterfaceType classOrInterfaceType) {
         SymbolReference<? extends ResolvedTypeDeclaration> ref = null;
-        if (classOrInterfaceType.toString().indexOf('.') > -1) {
-            ref = typeSolver.tryToSolveType(classOrInterfaceType.toString());
+        String typeName = classOrInterfaceType.getNameAsString();
+        if (typeName.indexOf('.') > -1) {
```
Comment (ftomassetti) [inline javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/declarations/JavaParserInterfaceDeclaration.java:314]: we should probably have some utility class for this sort of things

File: javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolver.java [hunk 3]
```diff
@@ -35,20 +35,26 @@
 import java.util.Optional;
 import java.util.concurrent.ExecutionException;
 
+import static com.github.javaparser.ParseStart.*;
+import static com.github.javaparser.ParserConfiguration.LanguageLevel.*;
 import static com.github.javaparser.Providers.provider;
 
 /**
+ * Defines a directory containig source code that should be used for solving symbols.
+ * The directory must correspond to the root package of the files within.
```
Comment (ftomassetti) [inline javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolver.java:46]: Yes, I should have written this a long time ago...

File: javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolver.java [hunk 4]
```diff
@@ -63,6 +69,11 @@ public JavaParserTypeSolver(File srcDir) {
             throw new IllegalStateException("SrcDir does not exist or is not a directory: " + srcDir.getAbsolutePath());
         }
         this.srcDir = srcDir;
+        javaParser = new JavaParser(
```
Comment (ftomassetti) [inline javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolver.java]: Should this be possibly configurable? Maybe we could optionally pass a parser configuration

File: javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolver.java [hunk 5]
```diff
@@ -142,25 +156,26 @@ public void setParent(TypeSolver parent) {
         String[] nameElements = name.split("\\.");
 
         for (int i = nameElements.length; i > 0; i--) {
-            String filePath = srcDir.getAbsolutePath();
+            StringBuilder filePath = new StringBuilder(srcDir.getAbsolutePath());
```
Comment (ftomassetti) [inline javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolver.java]: Better...

**Review by ftomassetti — APPROVED** [review]
Nice!

File: javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolver.java (hunk 5 above)
Comment (matozoid) [reply javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolver.java]: The yellow warnings of Intellij started to annoy me :-)

File: javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolver.java (hunk 4 above)
Comment (matozoid) [reply javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolver.java]: I just opened an issue for that. I could just as well fix it right away :-)

Comment (matozoid) [reply javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolver.java]: Fixed.

File: javaparser-core/src/main/java/com/github/javaparser/JavaParser.java (hunk 1 above)
Comment (matozoid) [reply javaparser-core/src/main/java/com/github/javaparser/JavaParser.java]: I rolled that back.

File: javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/declarations/JavaParserInterfaceDeclaration.java (hunk 2 above)
Comment (matozoid) [reply javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/declarations/JavaParserInterfaceDeclaration.java:314]: Well, what does it do?

Comment (ftomassetti) [reply javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/declarations/JavaParserInterfaceDeclaration.java:314]: I mean, to split a qualified name in its component or identify if a name is qualified

Comment (matozoid) [reply javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/declarations/JavaParserInterfaceDeclaration.java:314]: Oh, that actually sounds like `JavaParser.parseName(String)`

Comment (matozoid) [reply javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/declarations/JavaParserInterfaceDeclaration.java:314]: Oh, it's more complicated than that.

**Pull request merged**